pkg/autodiscovery/common/utils/kubelet.go

```diff
@@ -16,3 +20,35 @@ func GetCustomCheckID(annotations map[string]string, containerName string) (stri
 	id, found := annotations[fmt.Sprintf(checkIDAnnotationFormat, containerName)]
 	return id, found
 }
+
+// ValidateAnnotationsMatching detects if AD annotations don't match a valid container identifier
+func ValidateAnnotationsMatching(annotations map[string]string, containerIdentifiers map[string]bool, containerNames map[string]bool, adPrefix string) []error {
+	var errors []error
+	adAnnotation := fmt.Sprintf(`%s.+\..+`, adPrefix)
+	checkIDAnnotation := fmt.Sprintf(checkIDAnnotationFormat, ".+\\")
+
+	for annotation := range annotations {
+		if matched, _ := regexp.MatchString(checkIDAnnotation, annotation); matched {
+			// validate check.id annotation
+			err := validateIdentifier(annotation, containerNames, adPrefix)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		} else if matched, _ := regexp.MatchString(adAnnotation, annotation); matched {
+			// validate other AD annotations
+			err := validateIdentifier(annotation, containerIdentifiers, adPrefix)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+	return errors
+}
+
+func validateIdentifier(annotation string, containerIdentifiers map[string]bool, adPrefix string) error {
+	id := strings.Split(annotation[len(adPrefix):], ".")[0]
```

Comment:
This can crash if we got here by matching `checkIDAnnotation`, but `len(adPrefix) > len(annotation)`

Reply:
Ideally, `regexp.MatchString()` would only send annotations containing the `checkIDAnnotation`, which should always be longer than `adPrefix` since it contains the `adPrefix`. I'm going to try and make the regex more specific, but is there an example annotation I can look into that would cause the crash? I'm having trouble thinking of one 🙇

Reply:
Does `checkIDAnnotation` always contain `adPrefix`?